kmapper/visuals.py

```diff
@@ -476,20 +516,69 @@ def _format_tooltip(
         nbins,
     )
 
-    tooltip = env.get_template("cluster_tooltip.html").render(
-        projection_stats=projection_stats,
-        cluster_stats=cluster_stats,
-        custom_tooltips=custom_tooltips,
+    tooltip_data = {
+        'projection_stats': projection_stats,
+        'cluster_stats': cluster_stats,
+        'custom_tooltips': custom_tooltips,
+        'histogram': histogram,
+        'dist_label': "Member",
+        'node_id': node_ID,
+    }
+
+    return tooltip_data
+
+def render_d3_vis(
+    title,
+    mapper_summary,
+    histogram,
+    mapper_data,
+    colorscale
+    ):
+    # Find the module absolute path and locate templates
+    module_root = os.path.join(os.path.dirname(__file__), "templates")
+    env = Environment(loader=FileSystemLoader(module_root))
+
+    # Find the absolute module path and the static files
+    js_path = os.path.join(os.path.dirname(__file__), "static", "kmapper.js")
+    with open(js_path, "r") as f:
+        js_text = f.read()
+
+    css_path = os.path.join(os.path.dirname(__file__), "static", "style.css")
+    with open(css_path, "r") as f:
+        css_text = f.read()
+
+    if np.array(histogram).ndim == 1:
+        histogram = [histogram]
+
+    color_function_names = mapper_summary['custom_meta'].pop('color_function', None)
+    if isinstance(color_function_names, str):
+        color_function_names = [color_function_name]
+
+    if color_function_names is None:
+        raise Exception('''`color_function` required but not found in mapper_summary `custom_meta`''')
```

Comment:
Why would this not be found?  When would a user ever be able to hit this? Could you suggest a fix or place to read more in this exception text?

Reply:
visuals.py get really hairy because we have so many public functions, and the earlier one buries color_function_name within custom_meta, but only if it's passed -- here, https://github.com/scikit-tda/kepler-mapper/blob/1fdb48f6902bef6693b0dd54c8421abec26fdac2/kmapper/visuals.py#L218-L244

I'll think some more, but it's a parameter-threading mess right now, yeah.

Reply:
What do you think about https://github.com/scikit-tda/kepler-mapper/pull/196/commits/63bdf566d0865563bf81c8aeae2ec6ac00f81b98 ?

Reply:
🤕  I hear you and am definitely guilty of not keeping this simpler.  I'd be ok renaming some of these functions with a leading underscore.  Would moving them out of public space make it easier to move forward? I don't believe many of them are in the documentation.

Reply:
I'm for moving more out of public. Would need to check that doing so doesn't wreck plotlyviz.py, which relies on importing a lot of those functions rn.

Reply:
plotlyviz.py should still be able to use them. Biggest thing is adding a `_` prefix to indicate users should use them directly.

Reply:
done